kernel/src/engine/arrow_expression/evaluate_expression.rs

```diff
@@ -290,3 +319,223 @@ pub fn evaluate_predicate(
+/// of different types, they must be castable to a common supertype.
+/// If `result_type` is provided, it must match the common supertype derived by [`common_supertype`].
+pub fn coalesce_arrays(
+    arrays: &[ArrayRef],
+    result_type: Option<&DataType>,
+) -> Result<ArrayRef, ArrowError> {
+    if arrays.is_empty() {
+        return Err(ArrowError::InvalidArgumentError(
+            "Need at least one array".into(),
+        ));
+    }
+
+    // Early exit for single array case
+    if arrays.len() == 1 {
+        let array = &arrays[0];
+
+        // Validate result type if provided
+        if let Some(result_type) = result_type {
+            let result_arrow_type: ArrowDataType = result_type.try_into_arrow()?;
+            if array.data_type() != &result_arrow_type {
+                return Err(ArrowError::InvalidArgumentError(format!(
+                    "Single array type {:?} does not match requested result type {result_type:?}",
+                    array.data_type()
+                )));
+            }
+        }
+
+        return Ok(array.clone());
+    }
+
+    // Validate all arrays have the same length and collect types
+    let num_rows = arrays[0].len();
+    let mut types = Vec::with_capacity(arrays.len());
+    for (i, arr) in arrays.iter().enumerate() {
+        if arr.len() != num_rows {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Array at index {i} has length {}, expected {num_rows}",
+                arr.len()
+            )));
+        }
+        types.push(arr.data_type());
+    }
+
+    // Find common type
+    let supertype = common_supertype(&types).ok_or_else(|| {
+        ArrowError::InvalidArgumentError(format!(
+            "Could not find common supertype for types: {types:?}"
+        ))
+    })?;
+
+    // Validate result type if provided
+    if let Some(result_type) = result_type {
+        let result_arrow_type: ArrowDataType = result_type.try_into_arrow()?;
+        if result_arrow_type != supertype {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Requested result type {result_type:?} does not match common supertype {supertype:?}"
+            )));
+        }
+    }
+
+    // Cast arrays to common supertype and collect ArrayData for MutableArrayData
+    let (casted, array_data): (Vec<ArrayRef>, Vec<ArrayData>) = arrays
+        .iter()
+        .map(|arr| {
+            let casted_arr = if arr.data_type() == &supertype {
+                Ok(arr.clone())
+            } else {
+                cast(arr, &supertype)
+            };
+            casted_arr.map(|arr| {
+                let data = arr.to_data();
+                (arr, data)
+            })
+        })
+        .collect::<Result<Vec<_>, ArrowError>>()?
+        .into_iter()
+        .unzip();
+
+    // Build result
+    let mut mutable = MutableArrayData::new(array_data.iter().collect(), false, num_rows);
```

Comment:
`MutableArrayData::new()` requires a reference to owned `ArrayData`. I couldn't convince the compiler to do this in a more succinct (idiomatic?) way.
LMK if you see a better way.